pkg/virt-handler/heartbeat/heartbeat.go

```diff
@@ -0,0 +1,123 @@
+package heartbeat
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	k8scli "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/log"
+	virtutil "kubevirt.io/kubevirt/pkg/util"
+	clusterutils "kubevirt.io/kubevirt/pkg/util/cluster"
+	virtconfig "kubevirt.io/kubevirt/pkg/virt-config"
+	device_manager "kubevirt.io/kubevirt/pkg/virt-handler/device-manager"
+)
+
+type HeartBeat struct {
+	clientset               k8scli.CoreV1Interface
+	deviceManagerController device_manager.DeviceControllerInterface
+	clusterConfig           *virtconfig.ClusterConfig
+	host                    string
+	cpuManagerPath          string
+}
+
+func NewHeartBeat(clientset k8scli.CoreV1Interface, deviceManager device_manager.DeviceControllerInterface, clusterConfig *virtconfig.ClusterConfig, host string, clusterType clusterutils.ClusterType) *HeartBeat {
+	// This is a temporary workaround until k8s bug #66525 is resolved
+	var cpuManagerPath string
+	switch clusterType {
+	case clusterutils.ClusterTypeOCP3:
+		cpuManagerPath = virtutil.CPUManagerOS3Path
+	default:
+		cpuManagerPath = virtutil.CPUManagerPath
+	}
+	return &HeartBeat{
+		clientset:               clientset,
+		deviceManagerController: deviceManager,
+		clusterConfig:           clusterConfig,
+		host:                    host,
+		cpuManagerPath:          cpuManagerPath,
+	}
+}
+
+func (h *HeartBeat) Run(interval time.Duration, stopCh chan struct{}) {
+	go h.heartBeat(interval, stopCh)
+}
+
+func (h *HeartBeat) heartBeat(interval time.Duration, stopCh chan struct{}) {
+	// TODO Improve this
+	// Give the device plugins some time to initialize before going into the heartbeat loop
+	// The device plugins are normally almost immediately ready, still if we do the first heartbeat just
+	// a tiny bit before the device plugins are initialized, the node is not schedulable for a whole minute.
+	_ = wait.Poll(500*time.Millisecond, 5*time.Second, func() (done bool, err error) {
```

Comment:
Maybe would be good to define these intervals as constants somewhere ?

Reply:
I am uncertain if it is only used once in the code (one more indirection when looking at the code), but can do that.

Reply:
BTW, even if using once I think that such parameters should be extracted to global  constants for better clarity. This will also allow us to easily move parameters to external configuration files (the ones we think should be configurable)

Reply:
Done.